File: Data Structures & Algorithms/count-number-of-islands/submission-0.java
class Solution {
    int[][] directions = {{1,0}, {-1,0}, {0,1}, {0,-1}};
    public int numIslands(char[][] grid) {
        int row = grid.length;
        int col = grid[0].length;
        int count = 0;

        for(int i = 0; i < row; i++){
            for(int j = 0; j < col; j++){
                if(grid[i][j] == '1'){
                    dfs(grid, i, j);
                    count++;
                }
            }
        }
        return count;
    }

    private void dfs(char[][] grid, int r, int c){
        if(r < 0 || c < 0 || r >= grid.length || c >= grid[0].length){
            return;
        }

        if(grid[r][c] == '0') 
                return;

        grid[r][c] = '0';

        for(int[] dir : directions){
            dfs(grid, r + dir[0], c + dir[1]);
        }
    }
}

File: Data Structures & Algorithms/islands-and-treasure/submission-0.py
class Solution:
    def islandsAndTreasure(self, grid: List[List[int]]) -> None:
        direction = [[1, 0], [0, 1], [-1, 0], [0 ,-1]]
        q = deque()
        INF = 2147483647

        for r in range(len(grid)):
            for c in range(len(grid[0])):
                if grid[r][c] == 0:
                    q.append([r, c])
        

        while q:
            length = len(q)
            for i in range(length):
                r, c = q.popleft()
                for dr, dc in direction:
                    row, col = r + dr, c + dc
                    if (0 <= row < len(grid)
                    and 0 <= col < len(grid[0])
                    and grid[row][col] == INF):
                        grid[row][col] = grid[r][c] + 1
                        q.append([row, col])
        